File: ga/glm4_result_2024_6/result/1968D_cpp_glm/41_1968D.cpp
use std::io::{self, BufRead};

fn calculate(p: &Vec<usize>, a: &Vec<usize>, pos: usize, k: usize) -> i64 {
    let n = p.len();
    let mut mx = 0;
    let mut cur = 0;
    let mut vis = vec![false; n];
    while !vis[pos] && k > 0 {
        vis[pos] = true;
        mx = mx.max(cur + (k as i64) * (a[pos] as i64));
        cur += a[pos];
        pos = p[pos];
        k -= 1;
    }
    mx
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();

    let t: usize = lines.next().unwrap().unwrap().parse().unwrap();
    for _ in 0..t {
        let mut line = lines.next().unwrap().unwrap();
        let mut iter = line.split_whitespace();
        let n: usize = iter.next().unwrap().parse().unwrap();
        let k: usize = iter.next().unwrap().parse().unwrap();
        let ps1: usize = iter.next().unwrap().parse().unwrap();
        let ps2: usize = iter.next().unwrap().parse().unwrap();

        let mut p = Vec::with_capacity(n);
        let mut a = Vec::with_capacity(n);

        for _ in 0..n {
            p.push(iter.next().unwrap().parse::<usize>().unwrap() - 1);
        }

        for _ in 0..n {
            a.push(iter.next().unwrap().parse().unwrap());
        }

        let ans1 = calculate(&p, &a, ps1 - 1, k);
        let ans2 = calculate(&p, &a, ps2 - 1, k);

        if ans1 == ans2 {
            println!("Draw");
        } else if ans1 > ans2 {
            println!("Bodya");
        } else {
            println!("Sasha");
        }
    }
}

File: ga/4omini_result_2024_7/result/1968D_mini/41_1968D.rs
use std::io::{self, BufRead};

fn min(a: usize, b: usize) -> usize {
    if a < b { a } else { b }
}

fn maxl(a: i64, b: i64) -> i64 {
    if a > b { a } else { b }
}

fn main() {
    let stdin = io::stdin();
    let reader = stdin.lock();
    let mut lines = reader.lines();

    let t: usize = lines.next().unwrap().unwrap().parse().unwrap();
    
    for _ in 0..t {
        let line = lines.next().unwrap().unwrap();
        let mut inputs = line.split_whitespace().map(|x| x.parse::<i32>().unwrap());
        
        let n = inputs.next().unwrap() as usize;
        let k = inputs.next().unwrap() as usize;
        let mut a = inputs.next().unwrap() as usize;
        let mut b = inputs.next().unwrap() as usize;
        
        let m = min(n, k);
        let mut arr = vec![0; n + 1];
        let mut aa = vec![0; n + 1];
        
        let line = lines.next().unwrap().unwrap();
        for (i, value) in line.split_whitespace().enumerate() {
            arr[i + 1] = value.parse::<i32>().unwrap() as usize;
        }
        
        let line = lines.next().unwrap().unwrap();
        for (i, value) in line.split_whitespace().enumerate() {
            aa[i + 1] = value.parse::<i32>().unwrap();
        }
        
        let mut p1 = 0;
        let mut p2 = 0;
        let mut p1_ = 0;
        let mut p2_ = 0;

        for x in 1..=m {
            p1 = maxl(p1, (aa[a] as i64) * (k - x + 1) as i64 + p1_);
            p1_ += aa[a];
            a = arr[a];
            
            p2 = maxl(p2, (aa[b] as i64) * (k - x + 1) as i64 + p2_);
            p2_ += aa[b];
            b = arr[b];
        }
        
        if p1 > p2 {
            println!("Bodya");
        } else if p1 < p2 {
            println!("Sasha");
        } else {
            println!("Draw");
        }
    }
}